File: MatlabCode/06_b/DP_IC_Monte_Carlo.m
%% Monte Carlo simulation

clear all
close all;
clc;



%% Parameters

% Number of simulation runs
Nrun=1000; 

% Length of each run
T=100; 

% Load some parameters defined in file DP_IC_setup.m:
%   max capcity (C), demand prob (pW), stage cost coeff. (cost), 
%   max demand (W_max) and possible demand values (W_set)
[pW,cost,C,W_max,W_set]=DP_IC_setup();



%% Initialization

% Inizialize matrices to store results of simulation runs
x_star=zeros(T+1,Nrun);
u_star=zeros(T,Nrun);
gt_star=zeros(T,Nrun);
w_star=zeros(T,Nrun);

x_h1=zeros(T+1,Nrun);
u_h1=zeros(T,Nrun);
gt_h1=zeros(T,Nrun);
w_h1=zeros(T,Nrun);

x_h2=zeros(T+1,Nrun);
u_h2=zeros(T,Nrun);
gt_h2=zeros(T,Nrun);
w_h2=zeros(T,Nrun);



%% Compute different policies: Optimal, Heuristic 1, Heuristic 2

% Optimal Policy
[U_star, V] = DP_IC_optimal_policy(T);


% Heuristic policy 1: when level below max demand, refill inventory
u1=[C:-1:C-W_max+1];
u2=zeros(1,C-W_max+1);
U_h1 = repmat([u1 u2]',1,T);

% Heuristic policy 2: when level below max demand, order minimum quantity 
% to ensure the fulfillment of the demand 
u1=[W_max:-1:1];
u2=zeros(1,C-W_max+1);
U_h2 = repmat([u1 u2]',1,T);



%% Main loop

% Initial state: full
x0=C;

for k=1:Nrun
  % Optimal policy
  [x_star(:,k), u_star(:,k), gt_star(:,k), w_star(:,k)] = DP_IC_singlerun(T,U_star,x0); 
  
  % Heuristic policy 1
  [x_h1(:,k), u_h1(:,k), gt_h1(:,k), w_h1(:,k)] = DP_IC_singlerun(T,U_h1,x0); 

  % Heuristic policy 1
  [x_h2(:,k), u_h2(:,k), gt_h2(:,k), w_h2(:,k)] = DP_IC_singlerun(T,U_h2,x0); 

end



%% Plot and write results 

% Some plots
[J_star, J_h1, J_h2] = DP_IC_plot(gt_star, gt_h1, gt_h2, Nrun, T);

% Write some results
disp(' ');
disp(['Time horizon T = ' num2str(T) ]);
disp(['Number of simulation runs = ' num2str(Nrun)])
disp(['Initial condition x0 = ' num2str(x0)])
disp(' ');
disp(['Optimal policy mu(x) = [' num2str(U_star(:,1)') ']''']);
disp(' ');
disp('********* Average total cost ***********' )
disp(['Optimal policy     = ' num2str(mean(J_star))]); 
disp(['Heuristic policy 1 = ' num2str(mean(J_h1))]); 
disp(['Heuristic policy 2 = ' num2str(mean(J_h2))]); 
disp(' ');
disp(['Expected optimal total cost from x0: V0(x0) = ' num2str(sum(V(end,1))) ] );

File: PythonCode/06_b/DP_IC_Monte_Carlo.py
from utils import *


# parameters
Nrun = 1000     # number of simulations
T = 100         # time horizon of each simulation

# load parameters
pW, cost, C, W_max, W_set = DP_IC_setup()
N_state = C + 1

# init
x_star = np.zeros((T+1, Nrun), dtype=int)
u_star = np.zeros((T, Nrun), dtype=int)
gt_star = np.zeros((T, Nrun), dtype=int)
w_star = np.zeros((T, Nrun), dtype=int)

x_h1 = np.zeros((T+1, Nrun), dtype=int)
u_h1 = np.zeros((T, Nrun), dtype=int)
gt_h1 = np.zeros((T, Nrun), dtype=int)
w_h1 = np.zeros((T, Nrun), dtype=int)

x_h2 = np.zeros((T+1, Nrun), dtype=int)
u_h2 = np.zeros((T, Nrun), dtype=int)
gt_h2 = np.zeros((T, Nrun), dtype=int)
w_h2 = np.zeros((T, Nrun), dtype=int)

# compute different policies
# optimal policy
U_star, V = DP_IC_optimal_policy(T)

# heuristic policy #1: when level below max demand, refill
# U_h1 = [10, 10, 10, 10, ..., 10,
#          9,  9,  9,  9, ...,  9,
#          8,  8,  8,  8, ...,  8,
#          7,  7,  7,  7, ...,  7,
#          0,  0,  0,  0, ...,  0,
#          .,  .,  .,  ., ...,  .,
#          .,  .,  .,  ., ...,  .,
#          0,  0,  0,  0, ...,  0]

U_h1 = np.ones((N_state, Nrun), dtype=int)
for m, row in enumerate(U_h1):
    if m < W_max:
        U_h1[m] = row * (C - m)
    else:
        U_h1[m] = row * 0

# heuristic policy #2: when level below max demand, order minimum quantity to ensure fulfillment of demand
# U_h2 = [ 4,  4,  4,  4, ...,  4,
#          3,  3,  3,  3, ...,  3,
#          2,  2,  2,  2, ...,  2,
#          1,  1,  1,  1, ...,  1,
#          0,  0,  0,  0, ...,  0,
#          .,  .,  .,  ., ...,  .,
#          .,  .,  .,  ., ...,  .,
#          0,  0,  0,  0, ...,  0]

U_h2 = np.ones((N_state, Nrun), dtype=int)
for m, row in enumerate(U_h2):
    if m < W_max:
        U_h2[m] = row * (W_max - m)
    else:
        U_h2[m] = row * 0

# main loop
x0 = C

for k in range(Nrun):
    # optimal policy
    x_star[:, k], u_star[:, k], gt_star[:, k], w_star[:, k] = DP_IC_singlerun(T, U_star, x0)
    # heuristic 1
    x_h1[:, k], u_h1[:, k], gt_h1[:, k], w_h1[:, k] = DP_IC_singlerun(T, U_h1, x0)
    # heuristic 2
    x_h2[:, k], u_h2[:, k], gt_h2[:, k], w_h2[:, k] = DP_IC_singlerun(T, U_h2, x0)

# plot and write results
J_star, J_h1, J_h2 = DP_IC_plot(gt_star, gt_h1, gt_h2, Nrun, T)

print(f'Time horizon T = {T}')
print(f'Number of simulation runs = {Nrun}')
print(f'Initial condition x0 = {x0}')
print('')
print(f'Optimal policy mu(x) = {U_star[:, 1].T}')
print('-'*10 + 'Average Total Cost' + '-'*10)
print(f'Optimal policy: {np.mean(J_star)}')
print(f'Heuristic policy #1:  {np.mean(J_h1)}')
print(f'Heuristic policy #2:  {np.mean(J_h2)}')
